src/service/resource.js

```diff
@@ -365,19 +390,29 @@ export class Resource {
    * transitioned to the "ready for layout" state.
    */
   measure() {
+    let current = this;
+    let ancestor = this.getAmpAncestor();
     // Check if the element is ready to be measured.
-    // Placeholders are special. They are technically "owned" by parent AMP
-    // elements, sized by parents, but laid out independently. This means
-    // that placeholders need to at least wait until the parent element
-    // has been stubbed. We can tell whether the parent has been stubbed
-    // by whether a resource has been attached to it.
-    if (this.isPlaceholder_ &&
-        this.element.parentElement &&
-        // Use prefix to recognize AMP element. This is necessary because stub
-        // may not be attached yet.
-        startsWith(this.element.parentElement.tagName, 'AMP-') &&
-        !(RESOURCE_PROP_ in this.element.parentElement)) {
-      return;
+    while (ancestor) {
+      const resource = Resource.forElementOptional(ancestor);
+      // If there's an AMP ancestor that's still not stubbed (has a Resource),
+      // we don't know what to do yet.
+      if (!resource) {
+        return;
+      }
+      // If the current resource (not the ancestor) is a placeholder, allow it to
+      // measure. This is because placeholders are laid out independently of
+      // the ancestor element (even if they are sized and "owned" by it).
+      if (current.isPlaceholder_) {
+        break;
+      }
+      // If this ancestor isn't built yet, that means we don't know what kind
+      // of styles will be applied to this element. We must wait.
+      if (!ancestor.isBuilt()) {
```

Comment:
> With containers, we want to measure as soon as we can

Fixed.

>  We non-containers, we already all good, right? It will measure to 0 and then re-measure to the right value.

Unfortunately not. If the non-container isn't built yet, it will measure to 0 and never re-remeasure (because it thinks it has a proper measure). That's the [bug](https://github.com/ampproject/amphtml/issues/9267).